### ISSUE TITLE
Specifying markdown options in styles

### ISSUE THREAD

**Issue by halloleo** [msprev/panzer#15]
I normally use pandoc with

```
--read=markdown+autolink_bare_uris+multiline_tables+table_captions+definition_lists
```

How can I specify this in a style definition? According to the doco the option `read` is not allowed in the `commandline` yaml field.

**Comment by msprev** [comment]
Thanks! This was a deliberate design. Yaml headers are parsed by pandoc as markdown. The markdown reader options affect how the yaml fields get parsed, including any field that would set the markdown reader options. This can potentially create massive trouble. It is also unclear what the correct behaviour should be (read yaml data first with no markdown reader options? all option set?). There is also the possibility of triggering an unpredictable path down a series of reads, each of which reads a style that changes the reader, which requires reading everything again...

In short, you have to somehow tell panzer/pandoc the format of the incoming document + style definitions (including intended variant of markdown format). This cannot be specified in the style definitions as there is no correct way for panzer/pandoc to read those definitions without information about what format they are supposed to be in. Hope this makes sense!

**Comment by halloleo** [comment]
So does this mean I have to provide all the markdown extensions on the command line?

**Comment by msprev** [comment]
Yes, that is right. There is no other way for panzer/pandoc to know how it should interpret the document + style definitions.

**Comment by halloleo** [comment]
Sorry I’m not 100% with you. I do understand that there is a potential of trouble when the _main reader_ gets changed in the file, but the _extensions_ could be set through the yaml block, couldn’t they?

**Comment by msprev** [comment]
The problem is that there is no 'correct' way to read the values of fields in the yaml block these unless one makes an assumption about which version of markdown they should be read as, including which extensions are active. Some yaml metadata field values might even cause a syntax error in pandoc's parser if the appropriate extensions are not active. That is why the user has to specify extensions explicitly on the command line, along with the reader. If you can describe a correct behaviour here (panzer reads the style definition with no extensions? all extensions?), I'd consider implementing it.

There would be some performance penalty in an implementation. If markdown extensions invoked by style differ from the assumed default then everything would have to be re-read with the new extensions.

**Comment by halloleo** [comment]
I still do not see the universality of the problem:

Ok, if the "--read" option is set in the source document the assumptions about how yaml is represented in this document might cause trouble.

But if the "--read" option is set in a style file we do not have this problem: Styles files are by definition in the yaml text format which is not affected through the read option.

So maybe the "--read" option should be only forbidden in in-document style definitions, but not in separate yaml files?

**Comment by msprev** [comment]
> But if the "--read" option is set in a style file we do not have this
> problem: Styles files are by definition in the yaml text format which
> is not affected through the read option.

You have actually hit on the problem here. The style files are not just yaml format, they are yaml whose fields take values that are in _markdown_ format. panzer needs to know the value of these fields to take any action. But how can it do this without making an assumption about which markdown format they are in (i.e. which extensions should be switched on for the parser)?

> So maybe the "--read" option should be only forbidden in in-document style definitions, but not in separate yaml files?

Even if we set aside the previous worry, I don't think that I'd take this design decision. It is important that contents of style files are interchangeable with in-document definitions -- you should be able to cut and paste between them without errors. At the moment, they are parsed by the same engine.

**Comment by halloleo** [comment]
Hi Mark, thanks for explaining the logic behind panzer in so much patient detail. And sorry to be such a pain, but I have trouble to follow this logic:

You say the yaml syle files are parsed as what ever read-format is specified, correct? On the other side you do encourage the use of _global_ l style files (`styles.yaml` in the `.panzer` directory). 

So when I have markdown files which use a global style and I want to process them with panzer, the global style file is read as a markdown yaml file. Ok.

But I might have, say, rst files as well and they can use the global standard styles too. When I process these with panzer, the same style file now is read as a rst file which potential can cause lots of trouble. This means effectively that global styles can only used with _one_ format of input files.

Wouldn't it be better to say styles files are always read as, for example, markdown yaml and then they can be used for _all_ input formats. - And at the same time they then could contain `read` fields! :-)

**Comment by jgclark** [comment]
I realise this closed issue is related to my question the other day. Like @halloleo I don't follow all the logic of the problem. But I can see that the underscore characters in the extension names will require a clear parsing rule.

However, rather than parse as one long string, would it make it easier to force them to be specified like some of the other YAML? For example (using the original post, and adding one):

    extensions:
      autolink_bare_uris: true
      multiline_tables: true
      table_captions: true
      definition_lists: true
      smart: false

This would, ISTM, give a way of specifying both those to activate (in pandoc indicated by a +) and those ones on by default to suppress (indicated by a -).  Thanks.

**Comment by msprev** [comment]
This is complex to implement and it is not currently on the horizon. Pull requests are welcome.
